### PR TITLE
[docs] clarify that categorical features will be converted to integers internally

### LINKED ISSUE #4932: [questions] How to properly deal with categorical variables
According to official documents, the categorical features are better to be encoded as non-negative integers in LightGBM.(https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html). I encoded the categorical features as non-negative integers using OrdinalEncoder, but when I converted pandas dataframe to numpy array, the features will be converted to float. My question is could LGBM properly treat categorical features as float? What is the best way to deal with categorical features?

I am using sklearn API. The versions are python 3.8, LightGBM 3.3.1, pandas 1.1.5, numpy 1.19.5 and scikit-learn 1.0.1.

Here is a sample of my code. Thank you for help.

```
oe = preprocessing.OrdinalEncoder(dtype=int, handle_unknown='use_encoded_value', unknown_value=999)
feature_df[categorical_features] = oe.fit(feature_df[categorical_features].astype(str))

X = np.array(feature_df)
y = np.squeeze(np.array(target_df))
model.fit(X, y, sample_weight=weights,
                       categorical_feature=categorical_features,
                       feature_name=feature_features)
```

### PR DESCRIPTION
**Pull request by jmoralez**
Emphasizes that categorical features **will** be converted to `int` internally, to avoid the confusion that they should be provided as integers in the input data, as was asked in #4932. Also adds information on what will happen if negative values are found.

Here's how that section looks like:
image

Closes #4932.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
I think there are more places where this clarification should be made for the consistency in docs. See #1567 for example.

**Comment by shiyu1994** [comment]
Close and reopen to retrigger ci tests.

**Comment by StrikerRUS** [comment]
@jmoralez I'm very sorry, I just merged #5009 and only after doing that I noticed that these two PRs conflict with each other. Could you please resolve conflicts?

**Comment by jmoralez** [comment]
Haha no worries

**Comment by shiyu1994** [comment]
@jmoralez Thank you for doing this. Could you please rerun this script (https://github.com/microsoft/LightGBM/blob/master/helpers/parameter_generator.py)
so that the changes in `config.h` can be reflected on the documentation files.
Changes in the documentation files should also be updated to this branch to pass the static checks.

**Comment by jmoralez** [comment]
Thanks @shiyu1994. I didin't know I had to change the `config.h` file. Done in 349f797.

File: docs/Advanced-Topics.rst [hunk 1]
```diff
@@ -23,7 +23,8 @@ Categorical Feature Support
 -  Use ``categorical_feature`` to specify the categorical features.
    Refer to the parameter ``categorical_feature`` in `Parameters <./Parameters.rst#categorical_feature>`__.
 
--  Categorical features must be encoded as non-negative integers (``int``) less than ``Int32.MaxValue`` (2147483647).
+-  Categorical features will be cast to ``int`` so they must be encoded as non-negative integers (negative values will be treated as missing)
```
Comment (StrikerRUS) [inline docs/Advanced-Topics.rst:26]: For the consistency with other places.
```suggestion
-  Categorical features will be cast to ``int32`` so they must be encoded as non-negative integers (negative values will be treated as missing)
```

File: docs/Parameters.rst [hunk 2]
```diff
@@ -816,7 +816,7 @@ Dataset Parameters
 
    -  add a prefix ``name:`` for column name, e.g. ``categorical_feature=name:c1,c2,c3`` means c1, c2 and c3 are categorical features
 
-   -  **Note**: only supports categorical with ``int`` type (not applicable for data represented as pandas DataFrame in Python-package)
+   -  **Note**: all values will be cast to ``int`` (integer codes will be extracted from pandas categoricals in the Python-package)
```
Comment (StrikerRUS) [inline docs/Parameters.rst:819]:
```suggestion
   -  **Note**: all values will be cast to ``int32`` (integer codes will be extracted from pandas categoricals in the Python-package)
```

File: include/LightGBM/config.h [hunk 3]
```diff
@@ -697,7 +697,7 @@ struct Config {
   // desc = used to specify categorical features
   // desc = use number for index, e.g. ``categorical_feature=0,1,2`` means column\_0, column\_1 and column\_2 are categorical features
   // desc = add a prefix ``name:`` for column name, e.g. ``categorical_feature=name:c1,c2,c3`` means c1, c2 and c3 are categorical features
-  // desc = **Note**: only supports categorical with ``int`` type (not applicable for data represented as pandas DataFrame in Python-package)
+  // desc = **Note**: all values will be cast to ``int`` (integer codes will be extracted from pandas categoricals in the Python-package)
```
Comment (StrikerRUS) [inline include/LightGBM/config.h:700]:
```suggestion
  // desc = **Note**: all values will be cast to ``int32`` (integer codes will be extracted from pandas categoricals in the Python-package)
```

File: python-package/lightgbm/basic.py [hunk 4]
```diff
@@ -1155,7 +1155,7 @@ def __init__(self, data, label=None, reference=None,
             If list of int, interpreted as indices.
             If list of str, interpreted as feature names (need to specify ``feature_name`` as well).
             If 'auto' and data is pandas DataFrame, pandas unordered categorical columns are used.
-            All values in categorical features should be less than int32 max value (2147483647).
+            All values in categorical features will be cast to int32 and thus should be less than int32 max value (2147483647).
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:1158]: One more place in `refit()`:
https://github.com/microsoft/LightGBM/blob/9e73cee8aef2f5c5ea2d0f541abf3ca1d9cb832c/python-package/lightgbm/basic.py#L3563

File: python-package/lightgbm/engine.py [hunk 5]
```diff
@@ -460,7 +460,7 @@ def cv(params, train_set, num_boost_round=100,
         If list of int, interpreted as indices.
         If list of str, interpreted as feature names (need to specify ``feature_name`` as well).
         If 'auto' and data is pandas DataFrame, pandas unordered categorical columns are used.
-        All values in categorical features should be less than int32 max value (2147483647).
+        All values in categorical features will be cast to int32 and thus should be less than int32 max value (2147483647).
```
Comment (StrikerRUS) [inline python-package/lightgbm/engine.py:463]: One more place in `train()`:
https://github.com/microsoft/LightGBM/blob/717631af1836be9b03444ece485f3a3b5dbae5da/python-package/lightgbm/engine.py#L108

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
Please check my comments below:

**Review by StrikerRUS — APPROVED** [review]
Thanks for this clarification!

**Pull request merged**